### PR TITLE
Support for custom tls cipher suites in api server and kubelet

### LINKED ISSUE #41038: Disable SSL Medium Strength Cipher Suites Support.
**Is this a request for help?** (If yes, you should use our troubleshooting guide and community support channels, see http://kubernetes.io/docs/troubleshooting/.):Yes

**What keywords did you search in Kubernetes issues before filing this one?** (If you have found any duplicates, you should instead reply there.):Cipher,3DES,SHA1

---

**Is this a BUG REPORT or FEATURE REQUEST?** (choose one):
Disable SSL Medium Strength Cipher Suites Support.

**Kubernetes version** (use `kubectl version`):
Client Version: version.Info{Major:"1", Minor:"5", GitVersion:"v1.5.2", GitCommit:"08e099554f3c31f6e6f07b448ab3ed78d0520507", GitTreeState:"clean", BuildDate:"2017-01-12T04:57:25Z", GoVersion:"go1.7.4", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"5", GitVersion:"v1.5.2", GitCommit:"08e099554f3c31f6e6f07b448ab3ed78d0520507", GitTreeState:"clean", BuildDate:"2017-01-12T04:52:34Z", GoVersion:"go1.7.4", Compiler:"gc", Platform:"linux/amd64"}

**Environment**:
- **Cloud provider or hardware configuration**:
Bare-Metal

- **OS** (e.g. from /etc/os-release):
NAME="Ubuntu"
VERSION="16.04.1 LTS (Xenial Xerus)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 16.04.1 LTS"
VERSION_ID="16.04"
HOME_URL="http://www.ubuntu.com/"
SUPPORT_URL="http://help.ubuntu.com/"
BUG_REPORT_URL="http://bugs.launchpad.net/ubuntu/"
VERSION_CODENAME=xenial
UBUNTU_CODENAME=xenial

- **Kernel** (e.g. `uname -a`):
Linux xxx 4.4.0-62-generic #83-Ubuntu SMP Wed Jan 18 14:10:15 UTC 2017 x86_64 x86_64 x86_64 GNU/Linux

- **Install tools**:
- **Others**:

**What happened**:
Based on our security scanning: The remote service supports the use of medium strength SSL ciphers.
TLSv1
ECDHE-RSA-DES-CBC3-SHA Kx=ECDH Au=RSA Enc=3DES-CBC(168) Mac=SHA1 
DES-CBC3-SHA Kx=RSA Au=RSA Enc=3DES-CBC(168) Mac=SHA1

**What you expected to happen**:
Reconfigure the kube-apiserver to avoid use of medium strength ciphers.

**How to reproduce it** (as minimally and precisely as possible):

**Anything else we need to know**:

### PR DESCRIPTION
**Pull request by victorgp**
**What this PR does / why we need it**:
This pull request aims to solve the problem of users not able to set custom cipher suites in the api server.
Several users have requested this given that some default ciphers are vulnerable.
There is a discussion in #41038 of how to implement this. The options are:
- Setting a fixed list of ciphers, but users will have different requirements so a fixed list would be problematic.
- Letting the user set them by parameter, this requires adding a new parameter that could be pretty long with the list of all the ciphers.

I implemented the second option, if the ciphers are not passed by parameter, the Go default ones will be used (same behavior as now).

**Which issue this PR fixes** 
fixes #41038 

**Special notes for your reviewer**:
The ciphers in Go tls config are constants and the ones passed by parameters are a comma-separated list. I needed to create the `type CipherSuitesFlag` to support that conversion/mapping, because i couldn't find any way to do this type of reflection in Go.
If you think there is another way to implement this, let me know.

If you want to test it out, this is a ciphers combination i tested without the weak ones:

```
TLS_RSA_WITH_AES_128_CBC_SHA256,TLS_ECDHE_ECDSA_WITH_AES_128_CBC_SHA256,TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA256,TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256,TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256,TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384,TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384,TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305,TLS_ECDHE_ECDSA_WITH_CHACHA20_POLY1305,TLS_RSA_WITH_AES_128_CBC_SHA,TLS_RSA_WITH_AES_256_CBC_SHA,TLS_RSA_WITH_AES_128_GCM_SHA256,TLS_RSA_WITH_AES_256_GCM_SHA384,TLS_ECDHE_ECDSA_WITH_AES_128_CBC_SHA,TLS_ECDHE_ECDSA_WITH_AES_256_CBC_SHA,TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA,TLS_ECDHE_RSA_WITH_AES_256_CBC_SHA
```

If this is merged i will implement the same for the Kubelet.

**Release note**:
```release-note
kube-apiserver and kubelet now support customizing TLS ciphers via a `--tls-cipher-suites` flag
```

### CODE REVIEW

**Comment by victorgp** [comment]
CLA signed

File: staging/src/k8s.io/apiserver/pkg/util/flag/ciphersuites_flag.go
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/util/flag/ciphersuites_flag.go]: prefer tls constants here

File: staging/src/k8s.io/apiserver/pkg/server/options/serving.go
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/server/options/serving.go]: this should probably go in SecureServingOptions and be []string, tied to a string slice flag

File: staging/src/k8s.io/apiserver/pkg/util/flag/ciphersuites_flag.go
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/util/flag/ciphersuites_flag.go]: see https://github.com/liggitt/origin/blob/master/pkg/cmd/server/crypto/crypto.go#L72-L112 for an example, and https://github.com/liggitt/origin/blob/master/pkg/cmd/server/crypto/crypto_test.go#L16-L54 for a test to make sure the allowed values stays up to date with the ciphers supported by the TLS stack

**Comment by victorgp** [comment]
Parameter set to []string, new test added and using constants instead of numbers

File: staging/src/k8s.io/apiserver/pkg/server/options/serving.go
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/server/options/serving.go]: still needs to be moved to SecureServingOptions

File: staging/src/k8s.io/apiserver/pkg/server/options/serving.go
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/server/options/serving.go]: this can just be a util function, no need for a flag type now

**Comment by victorgp** [comment]
Sorry, i missed that part. Updated and type removed using just a function instead.

File: staging/src/k8s.io/apiserver/pkg/server/options/serving.go
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/server/options/serving.go]: omitted

File: staging/src/k8s.io/apiserver/pkg/util/flag/ciphersuites_flag.go
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/util/flag/ciphersuites_flag.go]: put these somewhere the apiserver and kubelet can both reference it...

**Comment by liggitt** [comment]
I'd like to see the option added to the kubelet at the same time... that'll help ensure the cipher util is put in a good package location

**Comment by victorgp** [comment]
Sure, i will implement the kubelet part too, i'll be out for 2 weeks though, so once i'm back.

**Comment by duglin** [comment]
@victorgp what's the status of the outstanding changes requested?

**Comment by victorgp** [comment]
@duglin sorry for the delay, most of this time i've been out and travelling for work so i barely had time for this. I will try to finish the kubelet part implementation soon and the other proposed changes.

**Comment by duglin** [comment]
@victorgp thanks for the update!

File: staging/src/k8s.io/apiserver/pkg/util/flag/ciphersuites_flag.go
Comment (victorgp) [reply staging/src/k8s.io/apiserver/pkg/util/flag/ciphersuites_flag.go]: After adding support for the kubelet, i didn't need to move this. It is accesible from the kubelet code, indeed, it was already imported so i didn't need to import anything.

Let me know if you still prefer to put it in another location, and if so, what would be the best one?

**Comment by victorgp** [comment]
I've implemented the same for the Kubelet

**Comment by victorgp** [comment]
I'm not sure if those build failed are expected, afaik Bazel builds are experimental.

Is there anything actually broken?

**Comment by duglin** [comment]
try the /retest cmd to see if it was just a burp

**Comment by victorgp** [comment]
After last changes, not failing tests anymore

File: staging/src/k8s.io/apiserver/pkg/util/flag/ciphersuites_flag.go
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/util/flag/ciphersuites_flag.go]: specifying an invalid cipher should return an error, not continue... with this logic, specifying completely invalid ciphers would fall back to default go ciphers

File: cmd/kubelet/app/server.go
Comment (liggitt) [inline cmd/kubelet/app/server.go]: handle error if invalid cipher is specified

File: cmd/kubelet/app/options/options.go
Comment (liggitt) [inline cmd/kubelet/app/options/options.go]: omitted

**Comment by liggitt** [comment]
a few nits. tests proving this is actually effective in setting the ciphers for the apiserver and kubelet would be good... things like this are easy to become accidentally disconnected in refactors if a test doesn't make sure it takes effect.

**Comment by victorgp** [comment]
Tests passing now with the cipher returning error if a cipher is not valid.

I will add more tests (integration) to test if the actual ciphers are working tonight or tomorrow

**Comment by duglin** [comment]
@liggitt any other comments/concerns?

File: test/integration/tls/ciphers_test.go
Comment (liggitt) [inline test/integration/tls/ciphers_test.go]: I'm confused why this is in a non-terminating for loop

File: pkg/kubelet/apis/kubeletconfig/v1alpha1/types.go
Comment (liggitt) [inline pkg/kubelet/apis/kubeletconfig/v1alpha1/types.go]: tlsCipherSuites?

**Comment by liggitt** [comment]
nit on config field name, and question about the integration test, lgtm otherwise.
cc @mtaufen

File: pkg/kubelet/apis/kubeletconfig/v1alpha1/types.go
Comment (mtaufen) [inline pkg/kubelet/apis/kubeletconfig/v1alpha1/types.go]: The json name should match the name on the config object, with the exception of the lowercase first letter. Please change the names on the config objects to `TLSCipherSuites`.

**Review by mtaufen — CHANGES_REQUESTED** [review]
Since this is just for TLS, I prefer we call the list of suites `TLSCipherSuites` everywhere, rather than just `CipherSuites`.

**Comment by victorgp** [comment]
All the tests passing, and all the changes requested done, let me know if anything is pending.

**Comment by liggitt** [comment]
/approve
thanks. go ahead and squash commits, then LGTM

**Review by mtaufen — APPROVED** [review]

**Comment by victorgp** [comment]
Squashed

**Comment by victorgp** [comment]
Any update on this? After all this time it would be nice to be merged before i need to rebase again ;)

**Comment by liggitt** [comment]
@kubernetes/sig-node-pr-reviews PTAL

**Comment by victorgp** [comment]
Tests passing again, you can now lgmt @liggitt  thanks

**Comment by duglin** [comment]
@liggitt any more comments on this one?

**Comment by liggitt** [comment]
none from me
/lgtm
/assign @derekwaynecarr 
for kubelet approval

**Comment by duglin** [comment]
@derekwaynecarr any comments/concerns?

**Comment by victorgp** [comment]
Is it normal that this is taking so much time to be reviewed?

**Comment by duglin** [comment]
@victorgp can you ask for that one test to be rerun?
@derekwaynecarr any chance of getting a final review?

**Comment by victorgp** [comment]
The tests are quite unstable, each time a different test, unrelated to this PR, fails.

**Comment by victorgp** [comment]
I had to make another commit to rebase and fix a test that was incompatible with current master

**Comment by victorgp** [comment]
I can't reproduce that unit test TestCRDDeletionCascading failing, and it seems is failing in other pull requests too, looks like it is flaky.
Like for example this simple one where only comments and a few bash modifications are added: https://github.com/kubernetes/kubernetes/pull/55976
Or this other one too: https://github.com/kubernetes/kubernetes/pull/56011

What's the process for flaky tests? shouldn't they be removed, better without them than having them creating noise

**Comment by victorgp** [comment]
Ok, it passed now...

**Comment by victorgp** [comment]
Another rebase, are you going to keep me rebasing this forever?? ;)

Please @derekwaynecarr @lavalamp @jlowdermilk can you take a look?

**Comment by victorgp** [comment]
@liggitt can someone else be assigned to also check this given that none of the assignees are responding?

**Comment by duglin** [comment]
now that kubecon is over, ping @liggitt for approval.

**Comment by victorgp** [comment]
Is there any chance the Kubernetes developers will consider this pull request? is there anything the people involved here that wants this security fix can do to have it merged?

**Comment by dims** [comment]
@dchen1107 @derekwaynecarr @Random-Liu @vishh @yujuhong @thockin  PTAL looks like this is ready!

/approve 
/lgtm

**Comment by swestcott** [comment]
@lavalamp are you available to review/approve this or delegate to someone else?

**Comment by victorgp** [comment]
Yet another rebase done.

I think @lavalamp is not paying attention to this PR, can anyone directly ping him somehow?

**Comment by swestcott** [comment]
> Yet another rebase done.
> I think @lavalamp is not paying attention to this PR, can anyone directly ping him somehow?

Pinging sig/node leads, @dchen1107, @derekwaynecarr, is there anyone else who's able to action the final approval on this?

**Comment by liggitt** [comment]
@swestcott I'll hunt down an approver, in the meantime the test failure is legitimate:

```
--- FAIL: TestKubeletConfigurationPathFields (0.00s)
	helpers_test.go:45: the following fields were expected, but missing from the result. If the field has been removed, please remove it from the kubeletConfigurationPathFieldPaths set and the KubeletConfigurationPathRefs function, or remove it from the kubeletConfigurationNonPathFieldPaths set, as appropriate:
		HostIPCSources[*]
		HostNetworkSources[*]
		HostPIDSources[*]
FAIL
```

looks like a faulty rebase on that test

File: staging/src/k8s.io/apiserver/pkg/server/config.go
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/server/config.go]: missed that this renamed... since this is already in the SecureServingInfo struct, no need to prefix with TLS. revert this rename

File: staging/src/k8s.io/apiserver/pkg/server/options/serving.go
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/server/options/serving.go]: same here, we're already in SecureServingOptions, so no need to prefix with TLS

**Comment by derekwaynecarr** [comment]
kubelet changes look good to me.

/approve

**Comment by liggitt** [comment]
@victorgp final stretch :)

fix the failing unit test, fixup the var names in the SecureServingOptions and SecureServingInfo structs, squash and we'll be good to go

**Comment by liggitt** [comment]
remove spurious `test/e2e/generated/bindata.go.tmp`?

**Comment by victorgp** [comment]
Done, names changed, test fixed and removed that test/e2e/generated/bindata.go.tmp that i don't know where it came from

**Comment by victorgp** [comment]
Ok, tests passing now.

Can someone merge?? ;)

**Comment by liggitt** [comment]
\o/

**Comment by duglin** [comment]
thanks @liggitt !

**Comment by victorgp** [comment]
Yeeees! Merged ;)

Thanks all!

File: cmd/kubelet/app/options/options.go
Comment (lavalamp) [inline cmd/kubelet/app/options/options.go]: List the values in this help message?

File: staging/src/k8s.io/apiserver/pkg/server/options/serving.go
Comment (lavalamp) [inline staging/src/k8s.io/apiserver/pkg/server/options/serving.go]: Same comment

File: staging/src/k8s.io/apiserver/pkg/util/flag/ciphersuites_flag.go
Comment (lavalamp) [inline staging/src/k8s.io/apiserver/pkg/util/flag/ciphersuites_flag.go]: If humans are supposed to type these things in, maybe omit the "TLS_" prefix, as it only adds verbosity?

Comment (liggitt) [reply staging/src/k8s.io/apiserver/pkg/util/flag/ciphersuites_flag.go]: those map to IANA names (https://www.iana.org/assignments/tls-parameters/tls-parameters.xml)... I'd keep them

File: cmd/kubelet/app/options/options.go
Comment (liggitt) [reply cmd/kubelet/app/options/options.go]: yeah, that's a good idea

File: test/integration/tls/ciphers_test.go
Comment (lavalamp) [inline test/integration/tls/ciphers_test.go]: I would prefer to see two tests for each cipher, one verifying it works when present, one verifying it doesn't when not.

**Comment by victorgp** [comment]
Ok will take care of these two things in another PR (the values in the help message and the tests)

**Comment by gregory-lyons** [comment]
Any chance of this getting cherry-picked to earlier versions?

**Comment by liggitt** [comment]
Given that it is a new feature and the size of the change, that is unlikely

File: cmd/kubelet/app/options/options.go
Comment (victorgp) [reply cmd/kubelet/app/options/options.go]: @lavalamp @liggitt i've made this PR to address this issue: https://github.com/kubernetes/kubernetes/pull/58920

**Comment by victorgp** [comment]
@lavalamp regarding the tests changes, i've might find an issue in how Go handles the ciphers.

From the list of Go supported 22 cipher suites here https://golang.org/pkg/crypto/tls/#pkg-constants there are only 10 that actually work and 12 don't work. Let me explain.

If you create a simple http server with the default Go TLS configuration, and you list the supported ciphers with a tool like openssl or nmap, only 10 show up.
If you do a HTTP request with any of the 12 not listed, the TLS negotiation fails, no matter what ciphers you select for your server, it fails either you select all of the possible 22 ciphers, or you select just the one the client is using, or you select none, in all cases, the TLS negotiation fails.

You just need to run this simple server:

```
package main

import (
    "net/http"
)
func main() {
    http.HandleFunc("/", nil) // set router
    http.ListenAndServeTLS(":9090", "/tmp/cert/apiserver.crt", "/tmp/cert/apiserver.key", nil) // set listen port
}
```

And use nmap to list the available ciphers:
```
$ nmap --script ssl-enum-ciphers -p 9090 localhost

Starting Nmap 7.01 ( https://nmap.org ) at 2018-01-27 14:06 CST
Nmap scan report for localhost (127.0.0.1)
Host is up (0.00013s latency).
PORT     STATE SERVICE
9090/tcp open  zeus-admin
| ssl-enum-ciphers: 
|   TLSv1.0: 
|     ciphers: 
|       TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA (secp256r1) - A
|       TLS_ECDHE_RSA_WITH_AES_256_CBC_SHA (secp256r1) - A
|       TLS_RSA_WITH_AES_128_CBC_SHA (rsa 2048) - A
|       TLS_RSA_WITH_AES_256_CBC_SHA (rsa 2048) - A
|       TLS_ECDHE_RSA_WITH_3DES_EDE_CBC_SHA (secp256r1) - C
|       TLS_RSA_WITH_3DES_EDE_CBC_SHA (rsa 2048) - C
|     compressors: 
|       NULL
|     cipher preference: server
|   TLSv1.1: 
|     ciphers: 
|       TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA (secp256r1) - A
|       TLS_ECDHE_RSA_WITH_AES_256_CBC_SHA (secp256r1) - A
|       TLS_RSA_WITH_AES_128_CBC_SHA (rsa 2048) - A
|       TLS_RSA_WITH_AES_256_CBC_SHA (rsa 2048) - A
|       TLS_ECDHE_RSA_WITH_3DES_EDE_CBC_SHA (secp256r1) - C
|       TLS_RSA_WITH_3DES_EDE_CBC_SHA (rsa 2048) - C
|     compressors: 
|       NULL
|     cipher preference: server
|   TLSv1.2: 
|     ciphers: 
|       TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256 (secp256r1) - A
|       TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384 (secp256r1) - A
|       TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA (secp256r1) - A
|       TLS_ECDHE_RSA_WITH_AES_256_CBC_SHA (secp256r1) - A
|       TLS_RSA_WITH_AES_128_GCM_SHA256 (rsa 2048) - A
|       TLS_RSA_WITH_AES_256_GCM_SHA384 (rsa 2048) - A
|       TLS_RSA_WITH_AES_128_CBC_SHA (rsa 2048) - A
|       TLS_RSA_WITH_AES_256_CBC_SHA (rsa 2048) - A
|       TLS_ECDHE_RSA_WITH_3DES_EDE_CBC_SHA (secp256r1) - C
|       TLS_RSA_WITH_3DES_EDE_CBC_SHA (rsa 2048) - C
|     compressors: 
|       NULL
|     cipher preference: server
|_  least strength: C

Nmap done: 1 IP address (1 host up) scanned in 1.00 seconds
```

The list of ciphers that down show up and never work are:
```
TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA256
TLS_RSA_WITH_RC4_128_SHA
TLS_ECDHE_ECDSA_WITH_AES_128_CBC_SHA
TLS_ECDHE_RSA_WITH_RC4_128_SHA
TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384
TLS_RSA_WITH_AES_128_CBC_SHA256
TLS_ECDHE_ECDSA_WITH_AES_128_CBC_SHA256
TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256
TLS_ECDHE_ECDSA_WITH_CHACHA20_POLY1305
TLS_ECDHE_ECDSA_WITH_RC4_128_SHA
TLS_ECDHE_ECDSA_WITH_AES_256_CBC_SHA
TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA
```

So, if you try a request using openssl lile:
`openssl s_client -cipher ECDH-RSA-RC4-SHA -connect localhost:9090`

You get the TLS negotiation and the simple http server prints
```
2018/01/27 15:52:15 http: TLS handshake error from 127.0.0.1:39312: tls: no cipher suite supported by both client and server
```

This means i can't implement those tests testing all the ciphers, because those 12, just don't work.

To me this is a bug in Go, unless i'm missing something.
Does any of you have an idea of what could be happening? Otherwise i will file a bug in Go

**Comment by liggitt** [comment]
> Does any of you have an idea of what could be happening? 

Some ciphers are only available when using serving certificates with certain characteristics.

**Comment by ralphbuk** [comment]
I know this is new function, but as this is security related it would be good if we could cherry pick to 1.8 & 1.9 so that we can let customers using general security scanner get rid of the security violations.

**Comment by krionbsd** [comment]
> I know this is new function, but as this is security related it would be good if we could cherry pick to 1.8 & 1.9 so that we can let customers using general security scanner get rid of the security violations.

Could you advice what k8s version should we use, which includes these TLS cipher fixes?

File: staging/src/k8s.io/apiserver/pkg/util/flag/ciphersuites_flag.go
Comment (neolit123) [reply staging/src/k8s.io/apiserver/pkg/util/flag/ciphersuites_flag.go]: looks like `TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305` and `TLS_ECDHE_ECDSA_WITH_CHACHA20_POLY1305` were added in golang lacking the hash portion of the suite IDs (`_SHA256`)

https://github.com/golang/go/issues/32061

but that's due to the same suffix was added last minute in the CHACHA20_POLY1305 draft:
https://tools.ietf.org/rfcdiff?url2=draft-ietf-tls-chacha20-poly1305-04.txt

:man_facepalming: 

xref https://github.com/kubernetes/kubernetes/issues/81145

**Pull request merged**
